$NetBSD: patch-ae,v 1.3 2007/09/22 04:20:01 bsadewitz Exp $

X-Git-Url: http://gitweb.freedesktop.org/?p=pixman.git;a=commitdiff;h=dc7c047d1c68f343b66e81d9e79084e4171e2634

[memlk] don't try to allocate new data for 0-rectangle regions
---

--- pixman/pixman-region.c.orig	2007-08-17 19:37:42.000000000 -0400
+++ pixman/pixman-region.c
@@ -2515,6 +2515,8 @@ pixman_region_init_rects (pixman_region1
 {
     int overlap;
 
+    /* if it's 1, then we just want to set the extents, so call
+     * the existing method. */
     if (count == 1) {
        pixman_region_init_rect(region,
                                boxes[0].x1,
@@ -2525,6 +2527,15 @@ pixman_region_init_rects (pixman_region1
     }
 
     pixman_region_init(region);
+
+    /* if it's 0, don't call pixman_rect_alloc -- 0 rectangles is
+     * a special case, and causing pixman_rect_alloc would cause
+     * us to leak memory (because the 0-rect case should be the
+     * static pixman_region_emptyData data).
+     */
+    if (count == 0)
+        return TRUE;
+
     if (!pixman_rect_alloc(region, count))
 	return FALSE;
 
